### PR TITLE
chore: Refactor Arrow Array and Schema allocation in ColumnReader and MetadataColumnReader

### LINKED ISSUE #1048: Refactor Arrow Array and Schema allocation in ColumnReader and MetadataColumnReader
### Describe the bug

We have refactored Arrow Array and Schema allocation for Comet execution before. Based on Arrow C Data interface spec, the two structures should be allocated by the consumer.

Another place we incorrectly allocate them is in ColumnReader and MetadataColumnReader. Previously the incorrect allocation caused a few strange errors. We should refactor the allocation in ColumnReader and MetadataColumnReader too.

### Steps to reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by viirya**
## Which issue does this PR close?

Closes #1048.

### CODE REVIEW

File: native/core/src/parquet/mod.rs [hunk 1]
```diff
@@ -52,7 +52,6 @@ const STR_CLASS_NAME: &str = "java/lang/String";
 /// Parquet read context maintained across multiple JNI calls.
 struct Context {
     pub column_reader: ColumnReader,
-    pub arrays: Option<(Arc<FFI_ArrowArray>, Arc<FFI_ArrowSchema>)>,
```
Comment (viirya) [inline native/core/src/parquet/mod.rs:55]: This refactoring simplifies the context and the logic. We don't need to keep the array and schema pointers in the producer side.

Comment (viirya) [reply native/core/src/parquet/mod.rs:55]: It is also easier to reason the array/schema release logic after this refactoring.

File: common/src/main/java/org/apache/comet/parquet/ColumnReader.java [hunk 2]
```diff
@@ -201,53 +207,56 @@ public CometDecodedVector loadVector() {
...
+    } else {
+      dictionary = new CometDictionary(dictionaryVector);
+    }
+
+    currentVector =
+        new CometDictionaryVector(
+            cometVector, dictionary, importer.getProvider(), useDecimal128, false, isUuid);
+
+    currentVector =
+        new CometDictionaryVector(cometVector, dictionary, importer.getProvider(), useDecimal128);
```
Comment (kazuyukitanimura) [inline common/src/main/java/org/apache/comet/parquet/ColumnReader.java:258]: Unrelated but we should not be needing to call `CometDictionaryVector` twice

File: native/core/src/parquet/mod.rs [hunk 3]
```diff
@@ -539,24 +537,16 @@ pub extern "system" fn Java_org_apache_comet_parquet_Native_currentBatch(
...
-    try_unwrap_or_throw(&e, |env| {
+    array_addr: jlong,
+    schema_addr: jlong,
+) {
+    try_unwrap_or_throw(&e, |_env| {
         let ctx = get_context(handle)?;
         let reader = &mut ctx.column_reader;
         let data = reader.current_batch();
-        let (array, schema) = data.to_spark()?;
+        data.move_to_spark(array_addr, schema_addr)?;
```
Comment (kazuyukitanimura) [inline native/core/src/parquet/mod.rs:547]: nit: maybe we can just return `data.move_to_spark(array_addr, schema_addr)` so that we do not need `ok(())`?

**Review by kazuyukitanimura — APPROVED** [review]
Looks good

File: native/core/src/parquet/mod.rs (hunk 3 above)
Comment (viirya) [reply native/core/src/parquet/mod.rs]: Okay.

**Comment by viirya** [comment]
Thanks @kazuyukitanimura

**Pull request merged**